### PR TITLE
*: Add security/authorization to etcd/client and etcdctl

### PR DESCRIPTION
**Pull request by barakmich**
All the variations work. This also adds the `-u` flag for all etcdctl actions, with proper password prompting, following the curl convention.

### CODE REVIEW

**Comment by xiang90** [comment]
I will test it manually first today or tomorrow (I am attending google io so there might be a review delay).

File: client/client.go [hunk 1]
```diff
@@ -293,6 +322,22 @@ func (c *simpleHTTPClient) Do(ctx context.Context, act httpAction) (*http.Respon
 	return resp, body, err
 }
 
+type authorizingHTTPClient struct {
```
Comment (xiang90) [inline client/client.go]: i cannot find where we used this. i think we are using the authedAction instead?

**Comment by xiang90** [comment]
`user add` has a unexpected behavior. 

I first create a root user. I expect to see a User root created infor. But it prints out an empty line.
Then I tried to create root again, I expected to see a user exists error. But this time I saw a created info.

```
Xiangs-Air:bin xiangli$ ./etcdctl user add root
New password: 

Xiangs-Air:bin xiangli$ ./etcdctl user list
root
Xiangs-Air:bin xiangli$ ./etcdctl user add root
New password: 
User root created
```

**Comment by xiang90** [comment]
`user passwd` creates a new user if the user did not exist. 

I tried to change the password of a non-existing user. I expect to see an error.
But it succeed with an empty line info and created a new user.

```
./etcdctl user passwd foo
New password: 

Xiangs-Air:bin xiangli$ ./etcdctl user list
foo
```

**Comment by xiang90** [comment]
`user remove` prints out `Key not found` if the user does not exist.

```
./etcdctl user remove not_exist
Key not found
```

**Comment by xiang90** [comment]
@barakmich How can I get the information of each user? It seems I can only list users in etcdctl.

**Comment by xiang90** [comment]
user revoke always return success even if the role does not exist

```
./etcdctl user revoke root -roles fleet
User root updated
./etcdctl user revoke root -roles fleet
User root updated
./etcdctl user revoke root -roles fleet
User root updated
```

**Comment by xiang90** [comment]
add an existing role returns unexpected error. We should print out something like `role xx already exists`

```
role add root
security: Role not updated. Use Grant/Revoke to update the role.
```

**Comment by xiang90** [comment]
@barakmich How to grant only read or write permission for a role?

**Comment by xiang90** [comment]
All key value operation needs to handle 502

```
./etcdctl get foo
Error:  502:  (unhandled http status [Unauthorized] with body [{"message":"Insufficient credentials"}]) [0]
```

**Comment by xiang90** [comment]
@barakmich These are all I have till now. The ctl is good in general.

**Comment by barakmich** [comment]
All addressed, depends on https://github.com/coreos/go-etcd/pull/219, and updating the Godep in this CL

**Comment by xiang90** [comment]
@yichengq Can you please also give this a try and review?

**Comment by yichengq** [comment]
Run some commands and find out:

1. `./bin/etcdctl user list` after enable security returns: `unexpected status code 401`

2. `./bin/etcdctl user passwd root` after enable security returns: `User root doesn't exist`

3. It is weird to see two root roles here:
```
unihorn@CoreOS ~/g/s/g/c/etcd> ./bin/etcdctl -u root:root user get root
User: root
Roles: root root
```

4. The line `password` and `new password` may confuse users:
```
unihorn@CoreOS ~/g/s/g/c/etcd> ./bin/etcdctl -u root user add etcd
Password:
New password:
Insufficient credentials
```

5. It is weird to see two guest roles here:
```
unihorn@CoreOS ~/g/s/g/c/etcd> ./bin/etcdctl -u root:root role list
guest
guest
root
```

6. maybe print out success when adding success?
```
unihorn@CoreOS ~/g/s/g/c/etcd> ./bin/etcdctl -u root:root role add fleet

```

7. i cannot grant role:
```
unihorn@CoreOS ~/g/s/g/c/etcd> ./bin/etcdctl -u root:root role get fleet
Role: fleet
KV Read:
KV Write:
unihorn@CoreOS ~/g/s/g/c/etcd> ./bin/etcdctl -u root:root role grant fleet -path "/fleet/*" -readwrite
Role fleet doesn't exist
```

Most of time it works really well and print out reasonable information. :)

**Comment by xiang90** [comment]
@barakmich 

We still need to solve 
```
./etcdctl user remove not_exist
Key not found
```

and (revoke should return an error if the role does not exist)
```
./etcdctl user revoke root -roles fleet
User root updated
./etcdctl user revoke root -roles fleet
User root updated
./etcdctl user revoke root -roles fleet
User root updated
```

**Comment by xiang90** [comment]
Do we need the `security-merging` prefix?

```
./etcdctl role grant foo -path "/fleet" -readwrite
security-merging: Granting duplicate read permission /fleet
```

**Comment by yichengq** [comment]
Except xiang's problems, the behavior looks good to me.

**Comment by philips** [comment]
Are there docs for these etcdctl features?

**Comment by barakmich** [comment]
Squashed, Rebased. Docs are now issue #2949

**Comment by xiang90** [comment]
Tested again. Works as expected.

**Comment by yichengq** [comment]
:ship:

**Comment by xiang90** [comment]
@barakmich The test failure has nothing to do with your pull request. You can merge it if you want.

**Pull request merged**